docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,401 @@
+# ADR 082: Data Companion API
+
+## Changelog
+
+- 2022-09-10: First draft (@thanethomson)
+
+## Status
+
+Accepted | Rejected | Deprecated | Superseded by
+
+## Context
+
+This ADR proposes to effectively offload some data storage responsibilities and
+functionality to **single external companion service** in order to:
+
+1. Improve core consensus stability.
+2. Eventually reduce the amount of functionality for which the core team is
+   responsible, thereby improving maintainability.
+3. Still cater for certain use cases that require this data and functionality.
+
+The way the system is currently built is such that a Tendermint node is mostly
+self-contained. While this philosophy initially allowed a certain degree of ease
+of node operation (i.e. simple UX), it has also lent itself to feature sprawl,
+with Tendermint being asked to take care of increasingly more than just
+consensus. Under the spotlight in this ADR are:
+
+1. **Event indexing**, which is required in order to facilitate arbitrary block
+   and transaction querying, as well as subscription for arbitrary events. We
+   have already seen, for instance, how the current (somewhat unreliable) event
+   subscription implementation can cause back-pressure into consensus and affect
+   IBC relayer stability (see [\#6729] and [\#7156]).
```

Comment:
Blocks and transactions are on-chain, Merkelized data from which events are derived. Querying block and transaction data doesn't really involve events at all. But this clause (and others) seem to imply some level of equivalence between these two things. Can you maybe speak a bit more on your understanding of events vs. other types of data?

Reply:
No, I don't see any equivalence between them. I picked those two kinds of data because I know they're important to different kinds of users.

They're also only examples of the kinds of data that we could expose via this interface.

Reply:
Purely in terms of information, and ignoring delivery mechanics, would these requirements not be satisfied by polling for the latest block in a hot loop?

Reply:
Yes, but with the recommended approach in this ADR Tendermint can automatically prune some of this data once it has assurances that the companion's received it. Consider the companion a "privileged" consumer of Tendermint data. Also, consider the possibility of this allowing us to offload responsibility for the majority of the RPC entirely to an external process that would store/index/present data and scale in whatever way users choose, independently of the Tendermint node.

Storage costs are incredibly high for operators at present, and if there's no need to store a particular kind of data to facilitate Tendermint's core functionality, why store it? Especially if someone wants to store/index it in a different way to how Tendermint natively stores/indexes it - in that case, operators would be paying those storage costs twice.

Reply:
> Consider the companion a "privileged" consumer of Tendermint data . . . the approach in this ADR [allows Tendermint to] automatically prune some of this data once it has assurances that the companion's received it.

A stream of events (as defined by this ADR) are derived from and (optimistically) describe on-chain state, but don't, and can't, provide, at a theoretical level, the consistency guarantees made by Tendermint via direct queries of that state, and so can't be treated as equivalent.

If the goal of the ADR is to reduce storage costs of Tendermint nodes, under the assumption that third party consumers will be reading and indexing data on a per-height basis, then a far more effective approach would be to introduce a manual per-height "purge" endpoint. (Maybe it even already exists! I don't know.) "Privileged" consumers of Tendermint data would poll the node for the latest height, fetch that height's block data through query RPCs that, unlike events streams, can be trusted to be authoritative, and then would call that "purge" endpoint when they've indexed and persisted the corresponding data.

Reply:
> a far more effective approach

Effective to what end? If we want to eventually offload responsibility entirely for providing an RPC interface to a separate process, and Tendermint only maintain the bare minimum set of RPC calls necessary for operational and transaction submission purposes, then the approach you're recommending is not effective.

If we wanted to continue having to maintain our own full RPC as we currently do, then your suggested approach could be effective.